projects/robots/pal_robotics/tiago++/controllers/tiago++/tiago++.c

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Description: Simple controller to make an "Hello" mouvement with both arms.
+ *              It is also possible to move the robot with the keyboard arrows.
+ */
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <webots/keyboard.h>
+#include <webots/motor.h>
+#include <webots/robot.h>
+
+#define MAX_SPEED 7.0  // [rad/s]
+#define N_PARTS 19
+#define MOTOR_LEFT N_PARTS - 2
+#define MOTOR_RIGHT N_PARTS - 1
+
+static WbDeviceTag robot_parts[N_PARTS];
+
+static void check_keyboard() {
+  double speeds_left = 0.0, speeds_right = 0.0;
+
+  int key = wb_keyboard_get_key();
+  if (key >= 0) {
+    switch (key) {
+      case WB_KEYBOARD_UP:
+        speeds_left = MAX_SPEED;
+        speeds_right = MAX_SPEED;
+        break;
+      case WB_KEYBOARD_DOWN:
+        speeds_left = -MAX_SPEED;
+        speeds_right = -MAX_SPEED;
+        break;
+      case WB_KEYBOARD_RIGHT:
+        speeds_left = MAX_SPEED;
+        speeds_right = -MAX_SPEED;
+        break;
+      case WB_KEYBOARD_LEFT:
+        speeds_left = -MAX_SPEED;
+        speeds_right = MAX_SPEED;
+        break;
+    }
+  }
+  wb_motor_set_velocity(robot_parts[MOTOR_LEFT], speeds_left);
+  wb_motor_set_velocity(robot_parts[MOTOR_RIGHT], speeds_right);
+}
+
+int main(int argc, char **argv) {
+  // init webots stuff
+  wb_robot_init();
+
+  const int time_step = wb_robot_get_basic_time_step();
+
+  // get devices
+  // initialize the robot's information
+  const char *names[N_PARTS] = {"head_2_joint",      "head_1_joint",      "torso_lift_joint",  "arm_right_1_joint",
+                                "arm_right_2_joint", "arm_right_3_joint", "arm_right_4_joint", "arm_right_5_joint",
+                                "arm_right_6_joint", "arm_right_7_joint", "arm_left_1_joint",  "arm_left_2_joint",
+                                "arm_left_3_joint",  "arm_left_4_joint",  "arm_left_5_joint",  "arm_left_6_joint",
+                                "arm_left_7_joint",  "wheel_left_joint",  "wheel_right_joint"};
+
+  double target_pos[N_PARTS] = {0.24, -0.67, 0.09, -0.43, -0.77, 0.00, 0.96, 1.41,     1.2,     0.00,
+                                0.74, -0.95, 0.06, 1.12,  1.45,  0.00, 0.00, INFINITY, INFINITY};
+
+  // configures and achieves the robot's position desired
+  for (int i = 0; i < N_PARTS; i++) {
+    robot_parts[i] = wb_robot_get_device(names[i]);
+    wb_motor_set_velocity(robot_parts[i], wb_motor_get_max_velocity(robot_parts[i]) / 2.0);
+    wb_motor_set_position(robot_parts[i], target_pos[i]);
+  }
+
+  // print user instructions
+  printf("\f");
+  printf("You can drive this robot:\n");
+  printf("Select the 3D window and use cursor keys:\n");
+  printf("\n");
```

Comment:
This message is not complete, it can be simplified with:

printf("You can drive this robot by selecting the 3D window and pressing the keyboard arrows.\n");

Same comment for the other controllers.